doc/tsm/tsi.md

```diff
@@ -0,0 +1,87 @@
+# TSI Proposal
+
+## Overview
+
+This is a proposal for a TSI (Time-Series Index) file format to provide fast, memory-mapped access to an inverted index of measurement, series and tag data. It operates at a per-shard level and can be appended to as new series are added.
+
+
+## Block types
+
+The file contains multiple block types:
+
+- Series Block - This block stores a series block ID and a list of series ids that map to series keys. The list is sorted by series key. The series id is only used internally within the index file and no guarantees are provided about preserving keys after a compaction.
+
+- Measurement Block - This block stores the measurement name and a list of tag keys. Each tag key points to a tag value block.
+
+- Tag Value Block - This block stores a list of values for a tag key that points to a sorted list of series ids. The series ids are used instead of series keys because the IDs are much smaller than the keys and sorted integer sets can be operated on more quickly.
+
+- WAL Block - This block stores a list of series keys that have not been compacted yet.
+
+## Usage
+
+Initially, series are appended to a WAL block and the series information is held in-memory. Once this reaches a threshold, the WAL block is compacted to a set of series, measurement, & tag value blocks. The WAL is discarded and these new blocks will serve as the index and will be operated upon through an mmap.
+
+Series keys can be queried by measurement and tag value by first jumping to the appropriate measurement block. References to measurement blocks will be held in-memory once the files are read in. From the measurement block, a binary search on the tag keys will allow you to find a particular key. Once the key is found, you can jump to the key's tag value block and perform a binary search to find a particular value and a list of associated series IDs.
```

Comment:
I have a question here about how often in real use-cases we find something in the index, versus not finding it in the index. This is a hard question to answer I guess...

If we spend a reasonable amount of time searching for a tag value that we won't ever find, I'm wondering if it's worth us using something like a Bloom Filter first, to determine if it's even worth using BS to look for a tag value. 

It would guarantee we wouldn't waste our time looking if the value didn't exist, and the false positive rate of ~1% would just mean doing extra, but unnecessary searches.